### PR TITLE
Handle edge cases in 'authorized'

### PR DESCRIPTION
**Pull request by noffle**
1. always recognize the original db author as authorized
2. recognize others as authorized when the only head is by the author in question

### CODE REVIEW

**Comment by noffle** [comment]
forgot to `standard`!

File: index.js [hunk 1]
```diff
@@ -258,23 +258,43 @@ HyperDB.prototype.authorized = function (key, cb) {
...
+
+  function tips (excludeKey, cb) {
+    var res = []
+    var pending = 0
+    var error
+    for (var i = 0; i < self._writers.length; i++) {
+      var writer = self._writers[i]
+      if (writer._feed.key.equals(excludeKey)) continue
+      pending++
+      writer.head(function (err, head) {
```
Comment (mafintosh) [inline index.js]: let's move this function out of the loop as `onhead` (has no loop state)

File: index.js [hunk 2]
```diff
@@ -258,23 +258,43 @@ HyperDB.prototype.authorized = function (key, cb) {
...
+      var feedKey = self.feeds[head.clock[j]].key
+      if (feedKey.equals(key)) {
+        return cb(null, true)
       }
     }
+
     return cb(null, false)
   })
+
+  function tips (excludeKey, cb) {
```
Comment (mafintosh) [inline index.js]: I think we should move tips to the prototype as _tips. wdyt?

File: index.js [hunk 3]
```diff
@@ -258,23 +258,43 @@ HyperDB.prototype.authorized = function (key, cb) {
   if (!cb) cb = noop
   var self = this
 
-  this.heads(function (err, heads) {
+  if (this.key.equals(key)) return process.nextTick(cb, null, true)
+
+  tips(key, function (err, heads) {
```
Comment (mafintosh) [inline index.js]: should call .ready before calling tips

File: index.js [hunk 4]
```diff
@@ -258,25 +258,56 @@ HyperDB.prototype.authorized = function (key, cb) {
...
 
-  this.heads(function (err, heads) {
+  if (this.key.equals(key)) return process.nextTick(cb, null, true)
+
+  this._tips(key, function (err, heads) {
     if (err) return cb(err)
+    var max = 0
     for (var i = 0; i < heads.length; i++) {
       var head = heads[i]
+      max = Math.max(head.clock.length)
```
Comment (ralphtheninja) [inline index.js]: `max = Math.max(head.clock.length, max)`?

Comment (mafintosh) [reply index.js]: classic @ralphtheninja , always lurking in the shadows

Comment (ralphtheninja) [reply index.js]: :grinning:

**Pull request merged**